### ISSUE TITLE
v1.26.3 breaks SSL support in Python 2.7.6

### ISSUE THREAD

**Issue by MadLittleMods** [urllib3/urllib3#2165]
### Subject

https://github.com/urllib3/urllib3/commit/6d607d400fdea62ced01a835486433392e3422fb (via https://github.com/urllib3/urllib3/pull/2095) introduces the usage of `hmac.compare_digest` which is only [supported in Python 2.7.7+](https://docs.python.org/dev/whatsnew/2.7.html#pep-466-network-security-enhancements-for-python-2-7). I don't see any note in the [changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst#1263-2021-01-26) about dropping support for Python 2.7.x or 2.7.6 specifically.

This results in the following error when downstream libraries use `urllib3` like [Ansible](https://github.com/ansible/ansible/blob/734c0535626ec8a452813c66787df34760615aa3/lib/ansible/module_utils/urls.py#L110-L129).

```
ERROR! 'module' object has no attribute 'compare_digest'
```

---

Related issue but different problem: https://github.com/urllib3/urllib3/issues/2051

Related issue: https://github.com/urllib3/urllib3/issues/2093

### Environment

```
('OS', 'Linux-3.13.0-170-generic-x86_64-with-Ubuntu-14.04-trusty')
('Python', '2.7.6')
('urllib3', '1.26.3')
```

<details>
<summary>Instructions for getting environment info</summary>

Describe your environment.
At least, paste here the output of:

```python
import platform
import urllib3

print("OS", platform.platform())
print("Python", platform.python_version())
print("urllib3", urllib3.__version__)
```

</details>

### Steps to Reproduce

```
$ python

>>> import sys
>>> sys.version
'2.7.6 (default, Nov 13 2018, 12:45:42) \n[GCC 4.8.4]'

>>> import hmac
>>> hmac.compare_digest("asdf", "asdf")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'compare_digest'
```

---

Run Ansible on a Ubuntu 14.04 (trusty) machine.
```
$ ansible-playbook playbook.yml
...
ERROR! 'module' object has no attribute 'compare_digest'
```

### Expected Behavior

No error is thrown.

### Actual Behavior

```
ERROR! 'module' object has no attribute 'compare_digest'
```

**Comment by sethmlarson** [comment]
[1.26.3 has backported support for `hmac.compare_digest`](https://github.com/urllib3/urllib3/blob/1.26.3/src/urllib3/util/ssl_.py#L30) so you shouldn't be seeing this error. This is very likely not related to the other issue you linked as https://github.com/urllib3/urllib3/issues/2051 is fixed and not related and https://github.com/urllib3/urllib3/issues/2093 only impacts v2 which isn't released yet.

Maybe paste the contents of our `urllib3/util/ssl_.py`? Might be getting modified by a third-party like a Linux repackager.

Closing this for now.

**Comment by MadLittleMods** [comment]
Thanks for the permalink to the tag itself! I was going off of `master` and comparing the dates to the tagged releases.

Testing this more directly, I can see that it is getting far enough after the `compare_digest` call so that it is actually running successfully :+1: -> [`src/urllib3/util/ssl_.py#L199-L202`](https://github.com/urllib3/urllib3/blob/361f1e2a61afdef86cb2feb0fa3f302e06c5fe2c/src/urllib3/util/ssl_.py#L199-L202). I should have done this before my original report.
```
$ python
>>> urllib3.util.ssl_.assert_fingerprint("edfd67a46a3f048fbd89763dce7eeffe", "edfd67a46a3f048fbd89763dce7eeffe")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/urllib3/util/ssl_.py", line 186, in assert_fingerprint
    .format(fingerprint, hexlify(cert_digest)))
urllib3.exceptions.SSLError: Fingerprints did not match. Expected "edfd67a46a3f048fbd89763dce7eeffe", got "1ed667d5f11f66b22c6a4e1dd3ac99ed".
```

I see the code has the backported `compare_digest` as well:

```
$ less /usr/local/lib/python2.7/dist-packages/urllib3/util/ssl_.py
...
_const_compare_digest = getattr(hmac, 'compare_digest', _const_compare_digest_backport)
...
if not _const_compare_digest(cert_digest, fingerprint_bytes):
```

---

I'm gonna look at all of the other `compare_digest` calls in the rest of the libraries to try to narrow it down.

Really appreciate the quick follow-up here ❤️

**Comment by MadLittleMods** [comment]
Looking into this further, I think I found the root cause.

I installed `urllib3[secure]` to try to get around the SNI errors so Ansible could `get_url` from an HTTPS website but the `secure` version of `urllib3` installs [`cryptography` as a requirement](https://github.com/urllib3/urllib3/blob/dc6284c80b5ae0f6724e91e6a356d27e8e3e6e1f/setup.py#L118).

When the `cryptography` library is available on your system, `ansible` tries to use that for the crypto backend instead of `pycrypto` which fails with Python 2.7.6 (latest available on Ubuntu 14 (trusty)).

To get around the issue, I installed `urllib3[secure]` around the area where I have to `get_url` and then uninstalled `cryptography` after that request.

---

For reference, `cryptography` uses `hmac.compare_digest(a, b)` at [`src/cryptography/hazmat/primitives/constant_time.py#L13`](https://github.com/pyca/cryptography/blob/e09cd90f77a31832bdde1d3652c115be282cced9/src/cryptography/hazmat/primitives/constant_time.py#L13)
